### PR TITLE
kubernetes: Config files + setup script for secure multiregion clusters

### PR DESCRIPTION
**Pull request by a-robinson**
tl;dr fill in a few constants at the top of secure.sh and then run it,
and you'll have a working secure multiregion cluster. Only tested on GKE
so far.

This relies on linking together each cluster's DNS servers so that
they're able to defer requests to each other for a special zone-scoped
namespace. This is a little hacky, but very maintainable and survivable
even if all the nodes in a cluster are taken down and brought back up
again later with different IP addresses.

The big caveat that I learned after I thought I was done, though, is
that GCE's internal load balancers only work within a region.
Unfortunately I had done my prototyping all within one region, with each
cluster just in a different zone. To get around this, I've had to switch
from exposing each DNS server on an internal IP to exposing them on
external IPs, which some users may not like, and which might tip the
scales in favor of a different solution. I'd be happy to discuss the
alternatives (hooking up a CoreDNS instance in each cluster to every
cluster's apiserver, or using istio multicluster) with anyone
interested.

I've only handled the secure cluster case in the script, but it can be
easily modified to also handle insecure clusters as an option. I should
translate the script to python for that, though. It's already more bash
than I'd want to ask people to use (and it requires bash v4, which macos
doesn't ship with by default).

Release note (general change): Configuration files and a setup script
for secure multiregion deployments on Kubernetes are now provided.

Fixes https://github.com/cockroachdb/cockroach/issues/25189

### CODE REVIEW

**Comment by kannanlakshmi** [comment]
Thanks Alex. Some comments that weren't intuitive to me though its more applicable for @jseldess as he puts together the docs for it: 

1. Would be good to specify the pre-req before starting this i.e., 2 or more K8s clusters in different zones or regions (maybe call out how zones/ regions work differently as you pointed out)
2. Setting up appropriate RBAC for each cluster - I struggled with this even though it seems obvious in retrospect 
3. How to delete everything (e.g., namespaces) and clean up folders when you're done
4. How to port forward to the web UI - I wasn't able to actually view it in the web UI though the CLI suggested port forwarding worked. Do you just need to port forward one of the nodes in any region? How does that work? e.g., `$ kubectl port-forward cockroachdb-0 8080 --context=gke_cockroach-workers_us-central1-a_lakshmi --namespace=us-central1-a`
image

**Comment by a-robinson** [comment]
Thanks for the feedback! I've converted everything to python and added a teardown.py script for cleaning up all the resources that setup.py creates. I'll work with @jseldess on documentation. I'm interested if @bobvawter has feedback on this before going too far with it, though, since the DNS linking is not ideal.

**Comment by a-robinson** [comment]
> How to port forward to the web UI - I wasn't able to actually view it in the web UI though the CLI suggested port forwarding worked. Do you just need to port forward one of the nodes in any region? How does that work? e.g., `$ kubectl port-forward cockroachdb-0 8080 --context=gke_cockroach-workers_us-central1-a_lakshmi --namespace=us-central1-a`

Yes, you should just have to port-forward to any of the nodes in any of the regions. I'm not sure what you mean by "I wasn't able to actually view it in the web UI".

**Comment by a-robinson** [comment]
ping @bobvawter?

**Review by mberhault — APPROVED** [review]
> Reviewed 7 of 10 files at r1, 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27092)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[cloud/kubernetes/multiregion/client-secure.yaml, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtZlqo6C3si6QwYnap:-LJtZlqp2PIj9mxQeLqw:b-ttijs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/client-secure.yaml#L11)):*
> ```YAML
>   containers:
>   - name: cockroachdb-client
>     image: cockroachdb/cockroach:v2.0.3
> ```

2.0.3? Here and elsewhere. We seem to be on 2.0.5 for the non-regional configs.
___
*[cloud/kubernetes/multiregion/cockroachdb-statefulset.yaml, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJt_UjVClRtyaAKTEra:-LJt_UjVClRtyaAKTErb:b-awypa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/cockroachdb-statefulset.yaml#L1)):*
> ```YAML
> apiVersion: v1
> ```

This is the insecure version, is it needed? All other configs (init, client) use secure mode.
___
*[cloud/kubernetes/multiregion/README.md, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtaLf7Df6PFKooKwE9:-LJtaLf7Df6PFKooKwEA:bu9od0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/README.md#L42)):*
> ```Markdown
> 
> This requirement is satisfied by clusters deployed in cloud environments such as Google Kubernetes Engine, and
> can also be satsified by on-prem environments depending on the [Kubernetes networking setup](https://kubernetes.io/docs/concepts/cluster-administration/networking/) used. If you want to test whether your cluster will work, you can run this basic network test:
> ```

s/satsified/satisfied/
___
*[cloud/kubernetes/multiregion/setup.py, line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtanru3ukIDSdNWXrY:-LJtanru3ukIDSdNWXrZ:b-es9pl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/setup.py#L9)):*
> ```Python
> 
> # Before running the script, fill in appropriate values for all the parameters
> # above the dashed line.
> ```

I think something that fails to run if not changed would be helpful. It would prevent people from just blindly running this.
___
*[cloud/kubernetes/multiregion/setup.py, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtazQx7I1A7wkhk10Y:-LJtazQx7I1A7wkhk10Z:b8u1r02) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/setup.py#L33)):*
> ```Python
> # Path to the cockroach binary on your local machine that you want to use
> # generate certificates. Defaults to trying to find cockroach in your PATH.
> # TODO: CHANGE BACK
> ```

Was that meant to be addressed?

**Review by a-robinson — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-:-LJtr73Q1Tt6zSB-8Pns:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[cloud/kubernetes/multiregion/client-secure.yaml, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtZlqo6C3si6QwYnap:-LJtj4hn-p3zyZ1UKxHU:blssda4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/client-secure.yaml#L11)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

2.0.3? Here and elsewhere. We seem to be on 2.0.5 for the non-regional configs.
</blockquote></details>

It's just fallen behind while waiting on review/feedback. Updated.
___
*[cloud/kubernetes/multiregion/cockroachdb-statefulset.yaml, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJt_UjVClRtyaAKTEra:-LJtjHEz2VdHKo35F57x:b4fw7z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/cockroachdb-statefulset.yaml#L1)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This is the insecure version, is it needed? All other configs (init, client) use secure mode.
</blockquote></details>

No, not needed. I don't think it's likely to be wanted by anyone. Removed.
___
*[cloud/kubernetes/multiregion/README.md, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtaLf7Df6PFKooKwE9:-LJtjVGr1EWPRiKY6SbD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/README.md#L42)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

s/satsified/satisfied/
</blockquote></details>

Done.
___
*[cloud/kubernetes/multiregion/setup.py, line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtanru3ukIDSdNWXrY:-LJtnV2v3NX-CxOCu1_i:b-yx9jum) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/setup.py#L9)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I think something that fails to run if not changed would be helpful. It would prevent people from just blindly running this.
</blockquote></details>

Done. I made the instructions clearer, kept a commented example, and added some basic input validation.
___
*[cloud/kubernetes/multiregion/setup.py, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27092#-LJtazQx7I1A7wkhk10Y:-LJtj_co901mGblwh4UN:b-ufbjmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d931015d03de6bc40e6913a9008f7a3e84f22d01/cloud/kubernetes/multiregion/setup.py#L33)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Was that meant to be addressed?
</blockquote></details>

Indeed, thanks.

**Comment by mberhault** [comment]
Nice! LGTM
